lib/Parse/ParseDecl.cpp

```diff
@@ -3900,6 +3900,15 @@ Parser::parseDecl(ParseDeclOptions Flags,
       const bool IsProbablyTupleDecl =
           Tok.is(tok::l_paren) && peekToken().isIdentifierOrUnderscore();
 
+      const bool IsProbablyStatement =
+          Tok.isIdentifierOrUnderscore() &&
+          peekToken().isAny(tok::period);
+
+      if (IsProbablyStatement) {
+        diagnose(Tok.getLoc(), diag::found_unexpected_stmt);
+        break;
+      }
```

Comment:
I'm a bit hesitant to approve this because:

1. We might end up emitting an incorrect "found an unexpected statement" even though we do not actually have a statement there. Perhaps the diagnostic should be worded so as to reflect the lower level of certainty? I'm not sure how to phrase that though. 🤔
2. This only triggers when we have a period following the identifier. Instead, if we think about it more holistically, there are several cases here:
    - We have a pattern like `ident :`,`ident =`,`ident ,` -> maybe it is a var decl.
    - We have a pattern like `ident (` -> maybe it is a func/operator decl with missing func
    - We have a pattern like `( ident` -> maybe we're trying to declare a tuple or something.
    - We have a pattern like `ident .` or `ident ?` or `ident [` or something else statement-like thing -> maybe we have a statement here (there's lots of possible statements, I don't know what kinds people accidentally end up writing here...). I think we should at least also check for `[`. As I wrote in the SR comments, it would be nice to actually just try parsing the thing and potentially treating it like a no-op in case that fails, but I don't know if that's doable...

Reply:
Agreed to both. That's why I left it as a draft but wanted to get it out there to have a pull request to look at. 

1. The diagnostic and the word `IsProbablyStatement` should be something else, maybe (until we get to 2. below) `IdentifierPeriod` or something plainly descriptive. 

2. For the holistic case though, I should (will) look at the grammar in TSPL and try to figure out what is a possible statement and maybe what else might appear here. The syntax is not valid here, but looks reasonable given the rest of how Swift works. Maybe `ResemblesStatement` or `IsStatementLike` might work as a description, with several cases that match common statement patterns. 

Does it make sense to try and fuzz it to see if we can generate the problem diagnostic cascade? I don't really know how the fuzzer works or if you can direct it to generate stuff this specific, but I can look.

Reply:
Looking at this again in the morning, it looks like it's expressions that are the problem. 

Statements can be a bunch of things that all start with keywords (`for`, `if` etc) or expressions. 

There are a lot of cases of expressions with identifiers at the beginning. The identifier at the beginning is the thing that is triggering the diagnostic cascade. In the `default:` case of the switch there's logic to guess it might be a `var` or `func` if the start is an identifier, but not to find the rest of the expressions that can appear there. 

I can replicate the cascade by putting `x + 3` in the body of a struct--a binary expression starting with an identifier. That was predictable using the thought process above--it's an expression starting with an identifier. I strongly suspect but have not yet tested that I can generate a bunch more of these from the grammar in TSPL. 

So, possibly the thing to do is to is, if the token is an identifier, try to parse it as an expression (using only the parse logic that accepts exprs starting with identifiers) then diagnose if that succeeds. That would handle a bunch of cases, if we can hook into the `ParseExpr` stuff without other problems. 

Another approach here is to check for the various expression patterns manually and generate the right diagnostic for them. That's not that hard, there are a finite number of them. it's just extending what I did for the `tok:period` case to other cases. That gives us better control in `ParseDecl` but might be kind of a drag to maintain as the language grows. 

For the diagnostic, it should probably be an error rather than note, and should have a different name because it's expressions that are causing the parse problem. I don't think you an do a fixit for it in the general case. I suppose you could say that these need to be assigned to something here and the fixit would be `x + 3` -> `let _ = x + 3` or something like that but I don't think that's helpful.

Reply:
Another component to this--the `IsProbablyFuncDecl` is only looking at whether the token is an identifier, not whether it's followed by an `l_paren`. That makes anything that starts with an identifier fall into the function class. Possibly just adding an `l_paren` check for the following token will bypass the cascade. I'm checking it now.

Reply:
adding an `l_paren` check to `IsProbablyFuncDecl` (and not classifying tokens without an `l_paren` as probably funcs) makes some tests fail, unsurprisingly, for instance if there's a generic parameter, so 

```
/Users/dfs/swift-source/swift/test/Parse/diagnostic_missing_func_keyword.swift:27:8: error: unexpected note produced: in declaration of 'Bar'
struct Bar {
       ^
/Users/dfs/swift-source/swift/test/Parse/diagnostic_missing_func_keyword.swift:30:42: error: expected error not produced
  %%<T: Brew> (lhs: T, rhs: T) -> T { // expected-error {{expected 'func' keyword in operator function declaration}} {{3-3=func }}
                                     ~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
                                     
/Users/dfs/swift-source/swift/test/Parse/diagnostic_missing_func_keyword.swift:31:42: error: expected error not produced
                                      // expected-error @-1 {{operator '%%' declared in type 'Bar' must be 'static'}}
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/dfs/swift-source/swift/test/Parse/diagnostic_missing_func_keyword.swift:32:63: error: incorrect message found
                                      // expected-error @-2 {{member operator '%%' must have at least one argument of type 'Bar'}}
                                                              ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
                                                              expected declaration
/Users/dfs/swift-source/swift/test/Parse/diagnostic_missing_func_keyword.swift:36:18: error: expected error not produced
  _: Int = 42 // expected-error {{expected 'var' keyword in property declaration}} {{3-3=var }}
             ~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
             
/Users/dfs/swift-source/swift/test/Parse/diagnostic_missing_func_keyword.swift:37:18: error: expected error not produced
              // expected-error @-1 {{property declaration does not bind any variables}}
~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/dfs/swift-source/swift/test/Parse/diagnostic_missing_func_keyword.swift:39:32: error: expected error not produced
  (light, dark) = (100, 200)// expected-error {{expected 'var' keyword in property declaration}} {{3-3=var }}
                            ~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
                            
/Users/dfs/swift-source/swift/test/Parse/diagnostic_missing_func_keyword.swift:41:16: error: expected error not produced
  a, b: Int // expected-error {{expected 'var' keyword in property declaration}} {{3-3=var }}
           ~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

so long story short that does not quite do it. For the first one `expected declaration` is kind of OK but for the second one it looks like we don't get an error at all. So this is not One Weird Trick to fix that up. Which is what I expected. 

going back to adding some individual statement syntax checks.

Reply:
Thinking about this a bit more: the original bug report was based on a particular usage pattern where someone was trying to set some property after initialization. That seems like a reasonable mistake to make. Handling all possible statements might be overkill, because it's unlikely that a developer will try out every possible statement unless that developer is a fuzzer.

I realize this is a bit of a 180 on my original stance, but maybe it's just fine if we have this special case? Say something like:

- We have a pattern like `ident :`,`ident =`,`ident ,` -> maybe it is a var decl.
- We have a pattern like `ident (` -> maybe it is a func/operator decl with missing func
- We have a pattern like `( ident` -> maybe we're trying to declare a tuple or something.
- We have a pattern like `ident` followed by something other than `:`/`=`/`,`/`(` -> maybe we have a statement.

This keeps the implementation simple. At the same time, maybe the diagnostic reflects some level of uncertainty "expected declaration but this looks like a statement". That seems a bit casual though. 😅

cc @brentdax: do you have suggestions on wording the diagnostic to reflect uncertainty without overly complicating the implementation to be 100% accurate? (For context, the main thing you might want to read is SR-13098 and this comment thread.)

Reply:
thanks Varun, that makes sense. If we limit it to the property-set after initialization (characterized by `id` `period`) then things are straightforward I think--we just catch that one case before we go to the existing missing-func logic. The existing missing-func logic will catch everything else that starts with an identifier, like the `x + 3` example, then cascade. It bothers me that that missing-func logic is catching stuff and generating inappropriate diagnostics but, hey. 

If we do a diagnostic like `error: properties cannot be set in the body of a struct; use the initializer` for the `id period` case, and let the other possible statements fall through to the possibly-func case, does that work?

Reply:
>  The existing missing-func logic will catch everything else that starts with an identifier, like the x + 3 example, then cascade. It bothers me that that missing-func logic is catching stuff and generating inappropriate diagnostics but, hey.

Right. That's why I suggested we also check that there is an opening paren after the identifier/operator, so we avoid the function mis-diagnosis.

> > We have a pattern like ident ( -> maybe it is a func/operator decl with missing func

Given that, it might be simpler to update the IsProbablyFuncDecl check to account for an extra opening paren and have the IsProbablyStatement go _after_ the func check (instead of before).

> If we do a diagnostic like error: properties cannot be set in the body of a struct; use the initializer for the id period case, and let the other possible statements fall through to the possibly-func case, does that work?

The problem is that we do not _know_ that the user is trying to set a property. It is likely, but we do not have scope information in the parser to actually check that (for good reasons). I would try to reduce the assumptions we're making. An inaccurate diagnostic reduces trust in the compiler and making more assumptions means that we are more likely to go wrong. So I would not phrase it in terms of a property.

Reply:
I agree with you. I might not have phrased it right. I'm sort of writing out thoughts and thinking them through and then revising them and some of the thoughts might have dropped out of the revisions :) . 

If we get past the `IsProbablyFuncDecl` logic, the diagnostic is just `expected declaration`, which is accurate but too terse, I think. 

one possibility is to tighten up the `IsProbablyFuncDecl` test, to only accept a left parenthesis or left angle bracket (for generics) after the identifier, then improve the `expected declaration` to something with more explanation like `expected declaration: struct/class bodies can only contain a declaration here.` What do you think?

Reply:
> one possibility is to tighten up the IsProbablyFuncDecl test, to only accept a left parenthesis or left angle bracket (for generics) after the identifier

Yes, that sounds right.

> then improve the expected declaration to something with more explanation like expected declaration: struct/class bodies can only contain a declaration here. What do you think?

I think we should try to avoid repeating information, especially in the same diagnostic. Instead, we can use `%select` to provide additional information when `IsPorbablyStmtDecl` is true (that we found something that looks like a statement).